iib/workers/tasks/build_merge_index_image.py

```diff
@@ -0,0 +1,278 @@
+    _get_external_arch_pull_spec,
+    _get_present_bundles,
+    _get_resolved_image,
+    _opm_index_add,
+    _prepare_request_for_build,
+    _push_image,
+    _update_index_image_build_state,
+    _update_index_image_pull_spec,
+    _verify_index_image,
+)
+from iib.workers.tasks.celery import app
+from iib.workers.tasks.utils import request_logger, run_cmd, set_registry_token
+
+
+__all__ = ['handle_merge_request']
+
+log = logging.getLogger(__name__)
+
+
+def _add_bundles_missing_in_source(
+    source_index_bundles,
+    target_index_bundles,
+    base_dir,
+    binary_image,
+    source_from_index,
+    request_id,
+    arch,
+    ocp_version,
+    overwrite_target_index_token=None,
+):
+    """
+    Rebuild index image with bundles missing from source image but present in target image.
+
+    If no bundles are missing in the source index image, the index image is still rebuilt
+    using the new binary image.
+
+    :param list source_index_bundles: bundles present in the source index image.
+    :param list target_index_bundles: bundles present in the target index image.
+    :param str base_dir: base directory where operation files will be located.
+    :param str binary_image: binary image to be used by the new index image.
+    :param str source_from_index: index image, whose data will be contained in the new index image.
+    :param int request_id: the ID of the IIB build request.
+    :param str arch: the architecture to build this image for.
+    :param str ocp_version: ocp version which will be added as a label to the image.
+    :param str overwrite_target_index_token: the token used for overwriting the input
+        ``source_from_index`` image. This is required for non-privileged users to use
+        ``overwrite_target_index``. The format of the token must be in the format "user:password".
+    :return: bundles which were added to the index image.
+    :rtype: list
+    """
+    set_request_state(request_id, 'in_progress', 'Adding bundles missing in source index image')
+    log.info('Adding bundles from target index image which are missing from source index image')
+    missing_bundles = []
+    source_bundle_digests = []
+    target_bundle_digests = []
+
+    for bundle in source_index_bundles:
+        if '@sha256:' in bundle['bundlePath']:
+            source_bundle_digests.append(bundle['bundlePath'].split('@sha256:')[-1])
+        else:
+            raise IIBError(
+                f'Bundle {bundle["bundlePath"]} in the source index image is not defined via digest'
+            )
+    for bundle in target_index_bundles:
+        if '@sha256:' in bundle['bundlePath']:
+            target_bundle_digests.append((bundle['bundlePath'].split('@sha256:')[-1], bundle))
+        else:
+            raise IIBError(
+                f'Bundle {bundle["bundlePath"]} in the target index image is not defined via digest'
+            )
+
+    for target_bundle_digest, bundle in target_bundle_digests:
```

Comment:
Instead of a for loop, this seems like a good candidate for the set union operation.